pkg/controller/controller.go

```diff
@@ -19,56 +33,595 @@ type Controller struct {
+	}
+	return err
+}
+
+func removeContainer(c []api.Container, name string) []api.Container {
+	for i, v := range c {
+		if v.Name == name {
+			c = append(c[:i], c[i+1:]...)
+			break
+		}
+	}
+	return c
+}
+func updateImageForRestikContainer(c []api.Container, name, image string) []api.Container {
+	for i, v := range c {
+		if v.Name == name {
+			c[i].Image = image
+			break
+		}
+	}
+	return c
+}
+
+func removeVolume(volumes []api.Volume, name string) []api.Volume {
+	for i, v := range volumes {
+		if v.Name == name {
+			volumes = append(volumes[:i], volumes[i+1:]...)
+			break
+		}
+	}
+	return volumes
+}
+
+func snapshotRetention(b *rapi.Backup) (string, error) {
+	cmd := fmt.Sprintf("/restic -r %s forget", b.Spec.Destination.Path)
+	if b.Spec.RetentionPolicy.KeepLastSnapshots > 0 {
+		cmd = fmt.Sprintf("%s --%s %d", cmd, rapi.KeepLast, b.Spec.RetentionPolicy.KeepLastSnapshots)
+	}
+	if b.Spec.RetentionPolicy.KeepHourlySnapshots > 0 {
+		cmd = fmt.Sprintf("%s --%s %d", cmd, rapi.KeepHourly, b.Spec.RetentionPolicy.KeepHourlySnapshots)
+	}
+	if b.Spec.RetentionPolicy.KeepDailySnapshots > 0 {
+		cmd = fmt.Sprintf("%s --%s %d", cmd, rapi.KeepDaily, b.Spec.RetentionPolicy.KeepDailySnapshots)
+	}
+	if b.Spec.RetentionPolicy.KeepWeeklySnapshots > 0 {
+		cmd = fmt.Sprintf("%s --%s %d", cmd, rapi.KeepWeekly, b.Spec.RetentionPolicy.KeepWeeklySnapshots)
+	}
+	if b.Spec.RetentionPolicy.KeepMonthlySnapshots > 0 {
+		cmd = fmt.Sprintf("%s --%s %d", cmd, rapi.KeepMonthly, b.Spec.RetentionPolicy.KeepMonthlySnapshots)
+	}
+	if b.Spec.RetentionPolicy.KeepYearlySnapshots > 0 {
+		cmd = fmt.Sprintf("%s --%s %d", cmd, rapi.KeepYearly, b.Spec.RetentionPolicy.KeepYearlySnapshots)
+	}
+	if len(b.Spec.RetentionPolicy.KeepTags) != 0 {
+		for _, t := range b.Spec.RetentionPolicy.KeepTags {
+			cmd = cmd + " --keep-tag " + t
+		}
+	}
+	if len(b.Spec.RetentionPolicy.RetainHostname) != 0 {
+		cmd = cmd + " --hostname " + b.Spec.RetentionPolicy.RetainHostname
+	}
+	if len(b.Spec.RetentionPolicy.RetainTags) != 0 {
+		for _, t := range b.Spec.RetentionPolicy.RetainTags {
+			cmd = cmd + " --tag " + t
+		}
+	}
+	output, err := execLocal(cmd)
+	return output, err
+}
+
+func execLocal(s string) (string, error) {
+	parts := strings.Fields(s)
+	head := parts[0]
+	parts = parts[1:]
+	cmdOut, err := exec.Command(head, parts...).Output()
+	return strings.TrimSuffix(string(cmdOut), "\n"), err
+}
+
+func getPasswordFromSecret(client *internalclientset.Clientset, secretName, namespace string) (string, error) {
+	secret, err := client.Core().Secrets(namespace).Get(secretName)
+	if err != nil {
+		return "", err
+	}
+	password, ok := secret.Data[Password]
+	if !ok {
+		return "", errors.New("Restic Password not found")
+	}
+	return string(password), nil
+}
+
+func (pl *Controller) addAnnotation(b *rapi.Backup) error {
+	annotation := make(map[string]string)
+	annotation[ImageAnnotation] = pl.Image
+	b.ObjectMeta.Annotations = annotation
```

Comment:
b is a pointer. so you are removing all the annotation that was set previously. instead make() only when they do not have any annotation or just add.